spectator-ext-jvm/src/main/java17/com/netflix/spectator/jvm/JavaFlightRecorder.java

```diff
@@ -0,0 +1,151 @@
+    Objects.requireNonNull(executor);
+    RecordingStream rs = new RecordingStream();
+    collectClassLoadingStatistics(registry, rs);
+    collectCompilerStatistics(registry, rs);
+    collectThreadStatistics(registry, rs);
+    collectVirtualThreadEvents(registry, rs);
+    collectZgcEvents(registry, rs);
+    executor.execute(rs::start);
+    return rs::close;
+  }
+
+  private static void collectClassLoadingStatistics(Registry registry, RecordingStream rs) {
+    AtomicLong prevLoadedClassCount = new AtomicLong();
+    AtomicLong prevUnloadedClassCount = new AtomicLong();
+    consume(ClassLoadingStatistics, rs, event -> {
+      long classesLoaded = event.getLong("loadedClassCount");
+      classesLoaded = classesLoaded - prevLoadedClassCount.getAndSet(classesLoaded);
+      registry.counter("jvm.classloading.classesLoaded").increment(classesLoaded);
+
+      long classesUnloaded = event.getLong("unloadedClassCount");
+      classesUnloaded = classesUnloaded - prevUnloadedClassCount.getAndSet(classesUnloaded);
+      registry.counter("jvm.classloading.classesUnloaded").increment(classesUnloaded);
+    });
+  }
+
+  private static void collectCompilerStatistics(Registry registry, RecordingStream rs) {
+    AtomicLong prevTotalTimeSpent = new AtomicLong();
+    consume(CompilerStatistics, rs, event -> {
+      long totalTimeSpent = event.getLong("totalTimeSpent");
+      totalTimeSpent = totalTimeSpent - prevTotalTimeSpent.getAndAdd(totalTimeSpent);
+      registry.counter("jvm.compilation.compilationTime").add(totalTimeSpent / 1000.0);
+    });
+  }
+
+  private static void collectThreadStatistics(Registry registry, RecordingStream rs) {
+    AtomicLong prevAccumulatedCount = new AtomicLong();
+    consume(JavaThreadStatistics, rs, event -> {
+      long activeCount = event.getLong("activeCount");
+      long daemonCount = event.getLong("daemonCount");
+      long nonDaemonCount = activeCount - daemonCount;
+      registry.gauge("jvm.thread.threadCount", "id", "non-daemon").set(nonDaemonCount);
+      registry.gauge("jvm.thread.threadCount", "id", "daemon").set(daemonCount);
+      long accumulatedCount = event.getLong("accumulatedCount");
+      long threadsStarted = accumulatedCount - prevAccumulatedCount.getAndSet(accumulatedCount);
+      registry.counter("jvm.thread.threadsStarted").increment(threadsStarted);
+    });
+  }
+
+  private static void collectVirtualThreadEvents(Registry registry, RecordingStream rs) {
+    consume(VirtualThreadPinned, rs, event ->
+      registry.timer("jvm.vt.pinned").record(event.getDuration())
+    ).withThreshold(Duration.ofMillis(20));
+    consume(VirtualThreadSubmitFailed, rs, event ->
+      registry.counter("jvm.vt.submitFailed").increment()
+    );
+  }
+
+  private static void collectZgcEvents(Registry registry, RecordingStream rs) {
+    consume(ZYoungGarbageCollection, rs, event ->
+      registry.timer("jvm.zgc.youngCollection", "tenuringThreshold", event.getString("tenuringThreshold"))
```

Comment:
`tenuringThreshold` appears to be a numeric value and if I understand correctly, ZGC can dynamically adjust that value. It would probably be better to have that as a separate metric rather than a dimension.

Reply:
Oh yeah, good point. Let me do that.